src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -62,11 +73,105 @@ private static IEnumerable<string> GetTagNames(Type tagHelperType)
                     name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
                 }
 
-                return new[] { ToHtmlCase(name) };
+                return new[]
+                {
+                    new ElementTarget
+                    {
+                        TagName = ToHtmlCase(name),
+                        AttributeNames = Enumerable.Empty<string>()
+                    }
+                };
             }
 
-            // Remove duplicate tag names.
-            return attributes.SelectMany(attribute => attribute.Tags).Distinct();
+            return targetElementAttributes.SelectMany(
+                targetElementAttribute =>
+                {
+                    IEnumerable<string> tagNames;
+                    IEnumerable<string> attributeNames = null;
+
+                    if (!TryGetValidatedNames(
+                            targetElementAttribute.Tags,
+                            pascalNameTarget: "Tag",
+                            errorSink: errorSink,
+                            names: out tagNames) ||
+                        (targetElementAttribute.Attributes != null &&
+                        !TryGetValidatedNames(
+                            targetElementAttribute.Attributes,
+                            pascalNameTarget: "Attribute",
+                            errorSink: errorSink,
+                            names: out attributeNames)))
+                    {
+                        return Enumerable.Empty<ElementTarget>();
+                    }
+
+                    return BuildElementTargets(tagNames, attributeNames, tagHelperType);
+                }).ToArray();
```

Comment:
use the `TagHelperDescriptorComparer` to eliminate duplicates here rather than when finding tag helpers that match start and end tags.